### ISSUE TITLE
[Question] is it possible for tmux to update a single pane in a window

### ISSUE THREAD

**Issue by nkh** [tmux/tmux#3136]
Hi,

I notice that, when a line changes, tmux updates the line in the current pane plus the continuation of the line in adjacent panes.
My guess is that tmux holds windows panes and puzzles the panes inside it.

This can clearly be seen when an image overlaying a pane is displayed and an adjacent pane text is updated. the image gets "overwritten" at the same height as the changed lines.

if necessary, I can film it.

Any solution for this?

**Comment by nicm** [comment]
tmux does not support images.

**Comment by nkh** [comment]
I understand that it doesn't but when an image is drawn in the framebuffer, updating a pane draws across the whole window; is there a way to have tmux only update the pane rather than the whole screen line?

**Comment by nkh** [comment]
Apropos not supporting images, wasn't there an attempt to do that via sixel? what happened to it?

**Comment by nicm** [comment]
No, tmux will update the terminal in the most appropriate way and if it doesn't know about images then it can't account for them.

I did quite a bit of work on SIXEL, up to the point where we needed a suitable way to convert a SIXEL image into ASCII (images need to degrade for non-SIXEL terminals, and probably under other circumstances since SIXEL is so slow). Someone else was going to take it further but for one reason or another never produced anything suitable for tmux and since I have limited interest in SIXEL it has not really progressed since then. If you want to carry on with it, there are two branches: `sixel-passthrough` which simply recognises SIXEL sequences and passes them on to the terminal - it would be quite easy to get this into a state for commit but it is of limited use - and `sixel` which has code to track the images and handle updating them when panes are moved and resized.

What is really needed is some suitably-licensed code to convert a SIXEL image into ASCII. It would be preferable if this was self-contained, free and understandable so it could be included with tmux itself, but it would be acceptable to write an internal API which wrapped a suitably-licensed external library (although with this SIXEL could not be enabled by default).

**Comment by nkh** [comment]
thank you for the answer, it's understandable that tmux doesn't handle things on the frame buffer; I only wondered if there was a way for tmux to update panes only rather than the whole line in the window (which I guess  is for efficiency sake).

I'll brush on my very limited sixel knowledge and I'll check the branches. If I understand you right, what's needed is to take a bunch of pixels in a sixel "cell" and convert that to a single ASCII character with the appropriate color code (supporting what color space?)?

**Comment by nkh** [comment]
libsixel is under MIT licensing and it lists all the components it uses and their license too. is that the type of license you are looking for?

https://saitoha.github.io/libsixel/

**Comment by nicm** [comment]
MIT is fine but we already have code for decoding sixel, what is needed is
a way to convert it to ascii, iirc that is the main thing although no doubt
there is other stuff in the branch that doesn't work correctly

**Comment by reportaman** [comment]
This is what I found for libs that convert images to ascii art:
https://en.wikipedia.org/wiki/Libcaca
https://en.wikipedia.org/wiki/AAlib
https://blog.filestack.com/api/an-api-to-convert-any-image-into-ascii/ (proprietary?)

**Comment by nkh** [comment]
when not using w3mimgpreview, that's integrated in my file manager, I use https://github.com/stefanhaustein/TerminalImageViewer, I like that it uses blocks, results are OK, single short 500 lines cpp file,  Apache-2.0 License, (https://github.com/stefanhaustein/TerminalImageViewer/blob/master/LICENSE).


FYI: I use Konsole as my terminal, used to have good reasons for it before I started using tmux, I just installed kitty and the problem I reported above is gone, no banding when displaying images with w3mimgdisplay.

What terminal with sixel support do you use to test? I can start by installing one.

**Comment by nkh** [comment]
if it fits the bil, I'll convert it.

**Comment by nicm** [comment]
I have merged master into the two sixel branches and here are some observations:

- `sixel-passthrough` should work, and TBH it could be merged - although it is a bit useless is may be more convenient for people using SIXEL. I am a bit wary of giving the impression tmux supports them though and carrying around this code if nobody is going to take it further.

- The `sixel` branch probably doesn't work, because I notice `tty_draw_images` is commented out - I guess it was broken in some past merge. So that at least would need to be fixed.

- The intent is that SIXELs are largely hidden behind an "image" API, so that we could potentially support other image formats, but this is not perfectly done (nor it necessarily need to be, it can be "good enough" now and improved if other image formats are added).

- I wrote SIXEL parsing and scaling code, it is in `sixel.c`. But I'm not tied to it particularly, if we want to throw it away and use a suitable external image library instead, that's fine. It should be hidden behind the image API and - importantly - any external dependencies must be a compile-time option.

- The current code is intended to replace SIXEL images by a plain box on terminals which do not support SIXEL (`sixel_to_screen`). This is what the SIXEL-to-ASCII code is intended to replace, but TBH it may be a better use of time to just stick with something simple like that and get the rest into shape (managing redrawing, moving and cropping images, making sure we don't draw over the top of them, etc). 

- I remember that SIXEL was very slow, partly that is because there is no caching and the entire image has to be sent to the terminal again every time we want to redraw it, but some may be the scaling/output code (`sixel_scale`/`sixel_print`), so that could probably do with being investigated.

**Comment by nkh** [comment]
fetched the branches and will look at the code and build the sixel-passthrough. I checked the docs to see how to run different servers and saw the -L option and the start-server command but I'm not sure how put everything together. I'd like to start a specific server for the sixel development  and point the tmux command to it; that would allow me to continue using the tmux version I have for other work and have a specific server+session for sixel, what's the incantation for that?

I'll have time to work more on this end of next week when the GF is on vacations but I may get back to you with some questions in the meantime.

I'll ask again, what did you use for sixel supporting terminal?

As for the sixel to ASCII code, I'll contact the tiv author to tell him that I'm re-writting it in C.

**Comment by nicm** [comment]
./tmux -Ltest new

will let you run it from the source tree with socket named test (or
anything you like).

I don't know what terminals support sixel apart from xterm.

**Comment by nkh** [comment]
tiv's code uses https://raw.githubusercontent.com/stefanhaustein/TerminalImageViewer/master/src/main/cpp/CImg.h which is has the following licensing:

 #                    CeCILL-C
 #                    The CeCILL-C license is close to the GNU LGPL.
 #                    ( http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html )
 #
 #                or  CeCILL v2.1
 #                    The CeCILL license is compatible with the GNU GPL.
 #                    ( http://www.cecill.info/licences/Licence_CeCILL_V2.1-en.html )
 
 not sure if that code will make it to the C version (in a form or another, it's cpp right now) but are those licenses OK for you?

**Comment by nicm** [comment]
no, I don't think so, sorry. i wouldn't want to bring in LGPL code and i
think like 10 pages of roll your own license is definitely out

you could ask if they will re/dual license under bsd, mit, apache?

doesn't libsixel have a convertor to ASCII? it is appropriately licensed
and we could just link against it as an external library

**Comment by nicm** [comment]
Or libcaca has a silly but free license and is able to load an image and
draw it as ASCII. Can we give it the image in some way and it return ASCII
instead of drawing it? Or fire callbacks to let tmux draw? I won't be back
at computer until Monday and it is hard to follow the API on my phone but
it seems worth a look.

**Comment by reportaman** [comment]
Perhaps unrelated... In my testing https://github.com/IonicaBizau/image-to-ascii based on  http://www.graphicsmagick.org produces best quality image to colored ascii art. Though feels a bit slow.

Here's an example: 
<img width="749" alt="Screen Shot 2022-04-02 at 12 38 55 AM" src="https://user-images.githubusercontent.com/71736629/161372497-df8866b5-bff2-467f-934b-ba462094ec0b.png">

**Comment by nkh** [comment]
it's certainly more "geeky" to display using ASCII7 than blocks.

To eliminate all licensing problems I think I'll write something from scratch. I'look into the caching and scaling, that's probably a good  ground for some speed optimization.

another "optimization" would be a multiple pass display, if tmux architecture would allow it, eg: display an image that gradually gets better.

**Comment by nicm** [comment]
OK bear in mind not all terminals support UTF-8. I would probably target
plain ASCII. I'm not sure about colour - you could probably use RGB and let
tmux convert down to 256/16/8 (it can already do this) or stick to the 256
colour palette.

**Comment by nkh** [comment]
Hi, as planned I'm going to get some time this week to start with this, I wonder if we shouldn't keep the conversation separate from this ticket by either opening a new one or if there's already one opened for sixe.

I'll answer my own and close this ticket in case someone has the same interest in mixing tmux and the frame buffer drawing.

**Comment by nicm** [comment]
Yes please open a new issue or a PR if you have further questions or code. Thanks!
